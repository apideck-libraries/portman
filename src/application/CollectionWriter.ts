import { CollectionDefinition } from 'postman-collection'
import {
  injectEnvVariables,
  orderCollectionRequests,
  overwriteCollectionKeyValues,
  overwriteCollectionValues,
  overwriteCollectionSecurityValues,
  writeCollectionPreRequestScripts,
  writeRawReplacements
} from '.'
import { GlobalConfig, PortmanConfig, PortmanOptions } from '../types'
import { isEmptyObject } from 'utils'

export class CollectionWriter {
  public collection: CollectionDefinition
  config: PortmanConfig
  options: PortmanOptions

  constructor(config: PortmanConfig, options: PortmanOptions, collection: CollectionDefinition) {
    this.config = config
    this.options = options
    this.collection = collection
  }

  public execute(): void {
    const { envFile, baseUrl } = this.options

    const { globals = {} } = this.config
    const {
      collectionPreRequestScripts = [],
      securityOverwrites = {},
      keyValueReplacements = {},
      valueReplacements = {},
      rawReplacements = [],
      orderOfOperations = []
    } = globals as GlobalConfig

    let collection = this.collection

    // --- Portman - Set Security values for Postman
    if (securityOverwrites && !isEmptyObject(securityOverwrites)) {
      collection = overwriteCollectionSecurityValues(collection, securityOverwrites)
    }

    // --- Portman - Search for keys in dictionary to set the value if key is found anywhere in collection
    if (keyValueReplacements && !isEmptyObject(keyValueReplacements)) {
      collection = overwriteCollectionKeyValues(collection, keyValueReplacements)
    }

    // --- Portman - Search for keys in dictionary to set the values if value is found anywhere in collection
    if (valueReplacements && !isEmptyObject(valueReplacements)) {
      collection = overwriteCollectionValues(collection, valueReplacements)
    }

    // --- Portman - Use .env to inject variables on Collection
    if (envFile) {
      collection = injectEnvVariables(collection, envFile, baseUrl)
    }

    // --- Portman - Set manually order Postman requests
    if (orderOfOperations && orderOfOperations.length > 0) {
      collection = orderCollectionRequests(collection, orderOfOperations)
    }

    // --- Portman - Set Postman pre-requests
    if (collectionPreRequestScripts && collectionPreRequestScripts.length > 0) {
      collection = writeCollectionPreRequestScripts(collection, collectionPreRequestScripts)
    }

    // --- Portman - Replace & clean-up Postman
    if (rawReplacements && rawReplacements.length > 0) {
      const collectionString = writeRawReplacements(
        JSON.stringify(collection, null, 2),
        rawReplacements
      )
      this.collection = JSON.parse(collectionString)
    } else {
      this.collection = collection
    }
  }
}
