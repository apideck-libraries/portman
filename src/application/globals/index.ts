export * from './injectEnvVariables'
export * from './orderCollectionRequests'
export * from './overwriteCollectionKeyValues'
export * from './overwriteCollectionSecurityValues'
export * from './overwriteCollectionValues'
export * from './renamePostmanCollection'
export * from './stripResponseExamples'
export * from './writeCollectionPreRequestScripts'
export * from './writeRawReplacements'
