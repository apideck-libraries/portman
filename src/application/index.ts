export * from './CollectionWriter'
export * from './globals'
export * from './newman'
export * from './overwrites'
export * from './tests'
export * from './TestSuite'
export * from './variables'
