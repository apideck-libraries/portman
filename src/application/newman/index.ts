export * from './runNewmanWith'
export * from './writeNewmanEnv'
