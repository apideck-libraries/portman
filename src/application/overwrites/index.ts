export * from './applyOverwrites'
export * from './overwriteRequestBody'
export * from './overwriteRequestHeaders'
export * from './overwriteRequestPathIdVariables'
export * from './overwriteRequestPathVariables'
export * from './overwriteRequestQueryParams'
export * from './overwriteRequestSecurity'
