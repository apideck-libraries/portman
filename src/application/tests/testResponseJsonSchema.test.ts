import { OpenAPIV3 } from 'openapi-types'
import { getOasMappedOperation } from '../../../__tests__/testUtils/getOasMappedOperation'
import { getPostmanMappedOperation } from '../../../__tests__/testUtils/getPostmanMappedOperation'
import { testResponseJsonSchema } from '../../application'
import { OasMappedOperation } from '../../oas'
import { PostmanMappedOperation } from '../../postman'
import { ContractTestConfig } from '../../types'

describe('testResponseJsonSchema', () => {
  let oasOperation: OasMappedOperation
  let pmOperation: PostmanMappedOperation

  beforeEach(async () => {
    oasOperation = await getOasMappedOperation()
    pmOperation = await getPostmanMappedOperation()
  })

  it('should add test for valid json schema', async () => {
    const schema = (oasOperation.schema?.responses?.['200'] as OpenAPIV3.ResponseObject)?.content
    pmOperation = testResponseJsonSchema(
      { enabled: true } as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should skip test for invalid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        foo: {
          $ref: '#/components/schemas/Foo'
        }
      }
    }

    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })
  it('should remove minItems on items levels, for valid json schema', async () => {
    const schema = {
      type: 'array',
      items: {
        required: ['name'],
        properties: {
          id: {
            type: 'integer',
            example: 10
          },
          name: {
            type: 'string',
            example: 'doggie'
          },
          status: {
            type: 'string',
            description: 'pet status in the store',
            enum: ['available']
          }
        },
        type: 'object'
      },
      minItems: 2
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove minItems on items levels, for types array/null, for valid json schema', async () => {
    const schema = {
      type: ['array', 'null'],
      items: {
        required: ['name'],
        properties: {
          id: {
            type: 'integer',
            example: 10
          },
          name: {
            type: 'string',
            example: 'doggie'
          },
          status: {
            type: 'string',
            description: 'pet status in the store',
            enum: ['available']
          }
        },
        type: 'object'
      },
      minItems: 2
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove maxItems on items levels, for valid json schema', async () => {
    const schema = {
      type: 'array',
      items: {
        required: ['name'],
        properties: {
          id: {
            type: 'integer',
            example: 10
          },
          name: {
            type: 'string',
            example: 'doggie'
          },
          status: {
            type: 'string',
            description: 'pet status in the store',
            enum: ['available']
          }
        },
        type: 'object'
      },
      maxItems: 2
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove maxItems on items levels, for types array/null, for valid json schema', async () => {
    const schema = {
      type: ['array', 'null'],
      items: {
        required: ['name'],
        properties: {
          id: {
            type: 'integer',
            example: 10
          },
          name: {
            type: 'string',
            example: 'doggie'
          },
          status: {
            type: 'string',
            description: 'pet status in the store',
            enum: ['available']
          }
        },
        type: 'object'
      },
      maxItems: 2
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove minItems on nested levels, for type array, for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: false
                },
                minItems: 2
              }
            },
            additionalProperties: false,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove minItems on nested levels, for types array/null, for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: ['array', 'null'],
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: false
                },
                minItems: 2
              }
            },
            additionalProperties: false,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove maxItems on nested levels, for type array, for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: false
                },
                maxItems: 2
              }
            },
            additionalProperties: false,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should remove maxItems on nested levels, for types array/null, for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: ['array', 'null'],
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: false
                },
                maxItems: 2
              }
            },
            additionalProperties: false,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      {} as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should toggle additional properties to false on nested levels for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: true
                },
                maxItems: 2
              }
            },
            additionalProperties: true,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      { additionalProperties: false } as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })

  it('should toggle additional properties to true on nested levels for valid json schema', async () => {
    const schema = {
      type: 'object',
      properties: {
        value: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              id: {
                type: 'string',
                example: 'p1'
              },
              name: {
                type: 'string',
                example: 'G-1'
              },
              ips: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    portMapping: {
                      type: 'integer',
                      example: 255
                    },
                    pool: {
                      type: 'string',
                      example: 'G-1'
                    }
                  },
                  additionalProperties: false
                },
                maxItems: 2
              }
            },
            additionalProperties: false,
            required: ['id', 'name']
          }
        }
      }
    }
    pmOperation = testResponseJsonSchema(
      { additionalProperties: true } as ContractTestConfig,
      schema,
      pmOperation,
      oasOperation
    )
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })
})
