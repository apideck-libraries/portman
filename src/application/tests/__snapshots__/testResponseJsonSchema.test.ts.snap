// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testResponseJsonSchema should add test for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"application/json\\":{\\"schema\\":{\\"x-graphql-type-name\\":\\"company\\",\\"type\\":\\"object\\",\\"required\\":[\\"status_code\\",\\"status\\",\\"service\\",\\"resource\\",\\"operation\\",\\"data\\"],\\"properties\\":{\\"status_code\\":{\\"type\\":\\"integer\\",\\"description\\":\\"HTTP Response Status Code\\",\\"example\\":200},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"HTTP Response Status\\",\\"example\\":\\"OK\\"},\\"service\\":{\\"type\\":\\"string\\",\\"description\\":\\"Apideck ID of service provider\\",\\"example\\":\\"zoho-crm\\"},\\"resource\\":{\\"type\\":\\"string\\",\\"description\\":\\"Unified API resource name\\",\\"example\\":\\"companies\\"},\\"operation\\":{\\"type\\":\\"string\\",\\"description\\":\\"Operation performed\\",\\"example\\":\\"one\\"},\\"data\\":{\\"type\\":\\"object\\",\\"required\\":[\\"name\\"],\\"additionalProperties\\":false,\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"readOnly\\":true,\\"example\\":\\"12345\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"Copper\\",\\"minLength\\":1},\\"interaction_count\\":{\\"type\\":[\\"integer\\",\\"null\\"],\\"example\\":1,\\"readOnly\\":true},\\"owner_id\\":{\\"type\\":\\"string\\",\\"example\\":\\"12345\\"},\\"image_url\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"https://www.spacex.com/static/images/share.jpg\\"},\\"description\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"A crm that works for you, so you can spend time on relationships instead of data.\\"},\\"vat_number\\":{\\"description\\":\\"VAT number\\",\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"BE0689615164\\"},\\"currency\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"USD\\"},\\"status\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"minLength\\":1,\\"example\\":\\"Open\\"},\\"fax\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"+12129876543\\"},\\"bank_accounts\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"iban\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"CH2989144532982975332\\"},\\"bic\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"AUDSCHGGXXX\\"}}}},\\"websites\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"required\\":[\\"url\\"],\\"properties\\":{\\"id\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"12345\\"},\\"url\\":{\\"type\\":\\"string\\",\\"example\\":\\"http://example.com\\",\\"minLength\\":1},\\"type\\":{\\"type\\":\\"string\\",\\"x-graphql-type-name\\":\\"WebsiteType\\",\\"enum\\":[\\"primary\\",\\"secondary\\",\\"work\\",\\"personal\\",\\"other\\"],\\"example\\":\\"primary\\"}}}},\\"addresses\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"123\\"},\\"type\\":{\\"type\\":\\"string\\",\\"x-graphql-type-name\\":\\"AddressType\\",\\"enum\\":[\\"primary\\",\\"secondary\\",\\"home\\",\\"office\\",\\"shipping\\",\\"billing\\",\\"other\\"],\\"example\\":\\"primary\\"},\\"name\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"HQ US\\"},\\"line1\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"Main street\\",\\"description\\":\\"Line 1 of the address e.g. number, street, suite, apt #, etc.\\"},\\"line2\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"apt #\\",\\"description\\":\\"Line 2 of the address\\"},\\"city\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"San Francisco\\",\\"description\\":\\"Name of city.\\"},\\"state\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"CA\\",\\"description\\":\\"Name of state\\"},\\"postal_code\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"94104\\",\\"description\\":\\"Zip code or equivalent.\\"},\\"country\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"US\\",\\"description\\":\\"country code according to ISO 3166-1 alpha-2.\\"},\\"latitude\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"40.759211\\"},\\"longitude\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"-73.984638\\"}}}},\\"social_links\\":{\\"type\\":\\"array\\",\\"items\\":{\\"required\\":[\\"url\\"],\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"12345\\"},\\"url\\":{\\"type\\":\\"string\\",\\"example\\":\\"https://www.twitter.com/apideck-io\\",\\"minLength\\":1},\\"type\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"twitter\\"}}}},\\"phone_numbers\\":{\\"type\\":\\"array\\",\\"items\\":{\\"required\\":[\\"number\\"],\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"12345\\"},\\"number\\":{\\"type\\":\\"string\\",\\"example\\":\\"111-111-1111\\",\\"minLength\\":1},\\"type\\":{\\"type\\":\\"string\\",\\"x-graphql-type-name\\":\\"PhoneType\\",\\"enum\\":[\\"primary\\",\\"secondary\\",\\"home\\",\\"office\\",\\"mobile\\",\\"assistant\\",\\"fax\\",\\"other\\"],\\"example\\":\\"primary\\"}}}},\\"emails\\":{\\"type\\":\\"array\\",\\"items\\":{\\"required\\":[\\"email\\"],\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"123\\"},\\"email\\":{\\"type\\":\\"string\\",\\"format\\":\\"email\\",\\"example\\":\\"elon@musk.com\\",\\"minLength\\":1},\\"type\\":{\\"type\\":\\"string\\",\\"x-graphql-type-name\\":\\"EmailType\\",\\"enum\\":[\\"primary\\",\\"secondary\\",\\"work\\",\\"personal\\",\\"billing\\",\\"other\\"],\\"example\\":\\"primary\\"}}}},\\"custom_fields\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"required\\":[\\"id\\"],\\"additionalProperties\\":false,\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"custom_technologies\\"},\\"value\\":{\\"anyOf\\":[{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"Uses Salesforce and Marketo\\"},{\\"type\\":[\\"number\\",\\"null\\"],\\"example\\":10},{\\"type\\":[\\"boolean\\",\\"null\\"],\\"example\\":true},{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"string\\"}}]}}}},\\"tags\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"string\\"},\\"example\\":[\\"New\\"]},\\"updated_by\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"12345\\",\\"readOnly\\":true},\\"created_by\\":{\\"type\\":[\\"string\\",\\"null\\"],\\"example\\":\\"12345\\",\\"readOnly\\":true},\\"updated_at\\":{\\"type\\":\\"string\\",\\"example\\":\\"2020-09-30T07:43:32.000Z\\",\\"readOnly\\":true},\\"created_at\\":{\\"type\\":\\"string\\",\\"example\\":\\"2020-09-30T07:43:32.000Z\\",\\"readOnly\\":true}}}}}}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove maxItems on items levels, for types array/null, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":[\\"array\\",\\"null\\"],\\"items\\":{\\"required\\":[\\"name\\"],\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"example\\":10},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"doggie\\"},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"pet status in the store\\",\\"enum\\":[\\"available\\"]}},\\"type\\":\\"object\\"}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove maxItems on items levels, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"array\\",\\"items\\":{\\"required\\":[\\"name\\"],\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"example\\":10},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"doggie\\"},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"pet status in the store\\",\\"enum\\":[\\"available\\"]}},\\"type\\":\\"object\\"}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove maxItems on nested levels, for type array, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":false}}},\\"additionalProperties\\":false,\\"required\\":[\\"id\\",\\"name\\"]}}}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove maxItems on nested levels, for types array/null, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":[\\"array\\",\\"null\\"],\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":false}}},\\"additionalProperties\\":false,\\"required\\":[\\"id\\",\\"name\\"]}}}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove minItems on items levels, for types array/null, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":[\\"array\\",\\"null\\"],\\"items\\":{\\"required\\":[\\"name\\"],\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"example\\":10},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"doggie\\"},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"pet status in the store\\",\\"enum\\":[\\"available\\"]}},\\"type\\":\\"object\\"}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove minItems on items levels, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"array\\",\\"items\\":{\\"required\\":[\\"name\\"],\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"example\\":10},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"doggie\\"},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"pet status in the store\\",\\"enum\\":[\\"available\\"]}},\\"type\\":\\"object\\"}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove minItems on nested levels, for type array, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":false}}},\\"additionalProperties\\":false,\\"required\\":[\\"id\\",\\"name\\"]}}}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should remove minItems on nested levels, for types array/null, for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":[\\"array\\",\\"null\\"],\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":false}}},\\"additionalProperties\\":false,\\"required\\":[\\"id\\",\\"name\\"]}}}}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should skip test for invalid json schema 1`] = `
Array [
  "// Response Validation Disabled due to Circular Reference
console.log('companiesOne response is not being validated against your spec!');
",
]
`;

exports[`testResponseJsonSchema should toggle additional properties to false on nested levels for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":false}}},\\"additionalProperties\\":false,\\"required\\":[\\"id\\",\\"name\\"]}}},\\"additionalProperties\\":false}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;

exports[`testResponseJsonSchema should toggle additional properties to true on nested levels for valid json schema 1`] = `
Array [
  "// Response Validation
const schema = {\\"type\\":\\"object\\",\\"properties\\":{\\"value\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\",\\"example\\":\\"p1\\"},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"},\\"ips\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"portMapping\\":{\\"type\\":\\"integer\\",\\"example\\":255},\\"pool\\":{\\"type\\":\\"string\\",\\"example\\":\\"G-1\\"}},\\"additionalProperties\\":true}}},\\"additionalProperties\\":true,\\"required\\":[\\"id\\",\\"name\\"]}}},\\"additionalProperties\\":true}

// Validate if response matches JSON schema 
pm.test(\\"[GET]::/crm/companies/:id - Schema is valid\\", function() {
    pm.response.to.have.jsonSchema(schema,{unknownFormats: [\\"int32\\", \\"int64\\"]});
});
",
]
`;
